File: SqlExpressions/Expressions/SqlQueryExpression.cs
using System.Collections.Generic;

namespace SqlExpressions.Expressions
{
    /// <summary>
    /// Class representing sql query
    /// </summary>
    /// <seealso cref="SqlExpressions.Expressions.SqlExpression" />
    public class SqlQueryExpression : SqlExpression
    {
        /// <summary>
        /// Gets the return type of the expression
        /// </summary>
        /// <value>
        /// The type of the return.
        /// </value>
        public override SqlType ReturnType => SqlType.Object;
        /// <summary>
        /// Gets the type of the expression.
        /// </summary>
        /// <value>
        /// The type of the expression.
        /// </value>
        public override SqlExpressionType ExpressionType => SqlExpressionType.Query;
        /// <summary>
        /// Accepts the specified visitor.
        /// </summary>
        /// <typeparam name="T">Return type</typeparam>
        /// <param name="visitor">The visitor.</param>
        /// <returns>Returns processed value from expression</returns>
        public override T Accept<T>(ISqlVisitor<T> visitor)
        {
            return visitor.Visit(this);
        }
        /// <summary>
        /// Selected expressions
        /// </summary>
        public IEnumerable<SqlExpression> Select { get; private set; }
        /// <summary>
        /// Source
        /// </summary>
        public SqlExpression From { get; private set; }
        /// <summary>
        /// Conditions
        /// </summary>
        public SqlExpression Where { get; private set; }
        /// <summary>
        /// Group clauses
        /// </summary>
        public IEnumerable<SqlExpression> GroupBy { get; private set; }
        /// <summary>
        /// Conditions on grouped
        /// </summary>
        public SqlExpression Having { get; private set; }
        /// <summary>
        /// Order by expressions
        /// </summary>
        public IEnumerable<SqlExpression> OrderBy { get; private set; }
        /// <summary>
        /// Additional annotations
        /// </summary>
        public IEnumerable<SqlExpression> Annotations { get; private set; }
        /// <summary>
        /// Limit query to first results
        /// </summary>
        public int Limit { get; private set; }

        /// <summary>
        /// Initializes a new instance of the <see cref="SqlQueryExpression"/> class.
        /// </summary>
        /// <param name="select">The select expressions.</param>
        /// <param name="from">Source expression.</param>
        /// <param name="where">Condition expression.</param>
        /// <param name="groupBy">Group by expressions.</param>
        /// <param name="having">Conditions on grouped.</param>
        /// <param name="orderBy">Order by expressions.</param>
        /// <param name="annotations">Additional annotations.</param>
        /// <param name="limit">The limit.</param>
        public SqlQueryExpression(
            IEnumerable<SqlExpression> @select,
            SqlExpression @from,
            SqlExpression @where = null,
            IEnumerable<SqlExpression> groupBy = null,
            SqlExpression having = null,
            IEnumerable<SqlExpression> orderBy = null,
            IEnumerable<SqlExpression> annotations = null,
            int limit = 0)
        {
            Select = @select;
            From = @from;
            Where = @where;
            GroupBy = groupBy;
            Having = having;
            OrderBy = orderBy;
            Annotations = annotations;
            Limit = limit;
        }
    }
}

File: Website/ClientApp/expression-tree/SqlQueryExpression.ts
import { SqlExpression, SqlExpressionBuilder } from "./SqlExpression";
import { SqlType } from "./SqlType";
import { SqlExpressionType } from "./SqlExpressionType";
import { SqlTableExpression } from "./SqlTableExpression";
import { TransferObjectId } from "./TransferObjectId";


export interface SqlQueryExpression extends SqlExpression {
    Annotations: SqlExpression[];
    From: SqlExpression;
    GroupBy: SqlExpression[];
    Having: SqlExpression;
    OrderBy: SqlExpression[];
    Select: SqlExpression[];
    Where: SqlExpression;
    Limit: number;
}

export class SqlQueryExpressionBuilder {
    Annotations: SqlExpression[];
    From: SqlExpression;
    GroupBy: SqlExpression[];
    Having: SqlExpression;
    OrderBy: SqlExpression[];
    Select: SqlExpression[];
    Where: SqlExpression;
    Limit: number;
    protected ExpressionType: SqlExpressionType;
    protected ReturnType: SqlType;


    setExpressionType(value: SqlExpressionType): SqlQueryExpressionBuilder {
        this.ExpressionType = value;
        return this;
    }

    setReturnType(value: SqlType): SqlQueryExpressionBuilder {
        this.ReturnType = value;
        return this;
    }

    setAnnotations(value: SqlExpression[]): SqlQueryExpressionBuilder {
        this.Annotations = value;
        return this;
    }

    setFrom(value: SqlExpression): SqlQueryExpressionBuilder {
        this.From = value;
        return this;
    }

    setGroupBy(value: SqlExpression[]): SqlQueryExpressionBuilder {
        this.GroupBy = value;
        return this;
    }

    setHaving(value: SqlExpression): SqlQueryExpressionBuilder {
        this.Having = value;
        return this;
    }

    setOrderBy(value: SqlExpression[]): SqlQueryExpressionBuilder {
        this.OrderBy = value;
        return this;
    }

    setSelect(value: SqlExpression[]): SqlQueryExpressionBuilder {
        this.Select = value;
        return this;
    }

    setWhere(value: SqlExpression): SqlQueryExpressionBuilder {
        this.Where = value;
        return this;
    }

    setLimit(value: number): SqlQueryExpressionBuilder {
        this.Limit = value;
        return this;
    }

    build(): SqlQueryExpression {
        return {
            ExpressionType: this.ExpressionType,
            ReturnType: this.ReturnType,
            Annotations: this.Annotations,
            From: this.From,
            GroupBy: this.GroupBy,
            Having: this.Having,
            OrderBy: this.OrderBy,
            Select: this.Select,
            Where: this.Where,
            Limit: this.Limit,
            TransferObjectId: TransferObjectId.SqlQueryExpression
        };
    }
}
